usr/src/lib/nsswitch/files/common/getnetgrent.c

```diff
@@ -0,0 +1,879 @@
+ *
+ *    -	keep a list of all the netgroups we haven't expanded yet (we just
+ *	want fast insertion and pop-first, so a linked list will do fine).
+ *	If we insert at the head, we get a depth-first search;  insertion
+ *	at the tail gives breadth-first (?), which seems preferable (?).
+ *
+ * A netgrnam struct contains pointers for both the hash-table and the list.
+ * It also contains the netgroup name;  note that we embed the name at the
+ * end of the structure rather than holding a pointer to yet another
+ * malloc()ed region.
+ *
+ * A netgrtab structure contains the hash-chain heads and the head/tail
+ * pointers for the expansion list.
+ */
+
+struct netgrnam {
+	struct netgrnam	*hash_chain;
+	struct netgrnam	*expand_next;
+	char		name[1];	/* Really [strlen(name) + 1] */
+};
+
+#define	HASHMOD	113
+
+struct netgrtab {
+	struct netgrnam	*expand_first;
+	struct netgrnam	**expand_lastp;
+	struct netgrnam	*hash_heads[HASHMOD];
+};
+
+static void
+ngt_init(struct netgrtab *ngt)
+{
+	(void) memset((void *)ngt, 0, sizeof (*ngt));
+	ngt->expand_lastp = &ngt->expand_first;
+}
+
+/* === ? Change ngt_init() and ngt_destroy() to malloc/free struct netgrtab */
+
+static void
+/* ==> ? Should return 'failed' (out-of-memory) status ? */
+ngt_insert(struct netgrtab *ngt, const char *name, size_t namelen)
+{
+	unsigned	hashval;
+	size_t		i;
+	struct netgrnam	*cur;
+	struct netgrnam	**head;
+
+#define	dummy		((struct netgrnam *)0)
+
+	if (__nss_files_netgr_debug != NULL)
+		__nss_files_netgr_debug(
+		    "ngt_insert: ngt=%p names=%s", ngt, name);
+
+	for (hashval = 0, i = 0;  i < namelen;  i++) {
+		hashval = (hashval << 2) + hashval +
+		    ((const unsigned char *)name)[i];
+	}
+	head = &ngt->hash_heads[hashval % HASHMOD];
+	for (cur = *head;  cur != 0;  cur = cur->hash_chain) {
+		if (strncmp(cur->name, name, namelen) == 0 &&
+		    cur->name[namelen] == 0) {
+			return;		/* Already in table, do nothing */
+		}
+	}
+	/* Create new netgrnam struct */
+	cur = (struct netgrnam *)
+	    malloc(namelen + 1 + (char *)&dummy->name[0] - (char *)dummy);
+	if (cur == 0) {
+		return;			/* Out of memory, too bad */
+	}
+	(void) memcpy(cur->name, name, namelen);
+	cur->name[namelen] = 0;
+
+	/* Insert in hash table */
+	cur->hash_chain = *head;
+	*head = cur;
+
+	/* Insert in expansion list (insert at end for breadth-first search */
+	cur->expand_next = 0;
+	*ngt->expand_lastp = cur;
+	ngt->expand_lastp = &cur->expand_next;
+
+#undef	dummy
+}
+
+static const char *
+ngt_next(struct netgrtab *ngt)
+{
+	struct netgrnam	*first;
+
+	if ((first = ngt->expand_first) == 0) {
+		return (0);
+	}
+	if ((ngt->expand_first = first->expand_next) == 0) {
+		ngt->expand_lastp = &ngt->expand_first;
+	}
+	return (first->name);
+}
+
+static void
+ngt_destroy(struct netgrtab *ngt)
+{
+	struct netgrnam	*cur;
+	struct netgrnam *next;
+	int		i;
+
+	for (i = 0;  i < HASHMOD;  i++) {
+		for (cur = ngt->hash_heads[i];  cur != 0; /* cstyle */) {
```

Comment:
NULL and try dropping the cstyle comment. If that causes a cstyle warning, I'd switch to a while loop.